desc/compute/utils.py

```diff
@@ -1347,3 +1360,776 @@ def body(i, mins):
+        A maximum of three bounce points can be detected in between knots.
+        The accuracy of the quadrature will increase as some function of
+        the number of knots over the number of detected bounce points.
+        So for well-behaved magnetic fields increasing resolution should increase
+        the accuracy of the quadrature.
+    method : str
+        The method to evaluate the integral.
+        The "direct" method exactly integrates a cubic spline of the integrand.
+        The other methods perform a Gauss quadrature and use independent cubic splines
+        for components in the integrand so that the singularity near the bounce points
+        can be captured more accurately than can be represented by a polynomial.
+
+    Returns
+    -------
+    bi : callable
+        This callable method computes the bounce integral F_ℓ(λ) for every
+        specified field line ℓ (constant rho and alpha), for every λ value in
+        ``pitch``.
+
+    Examples
+    --------
+    .. code-block:: python
+
+        bi = bounce_integral(eq)
+        F = bi(name, pitch)
+
+    """
+
+    def _direct(name, pitch=None):
+        """Compute the bounce integral of the named quantity.
+
+        Parameters
+        ----------
+        name : ndarray
+            Name of quantity in ``data_index`` to compute the bounce integral of.
+        pitch : ndarray
+            λ values to evaluate the bounce integral at.
+            If None, uses the values given to the parent function.
+
+        Returns
+        -------
+        F : ndarray, shape(pitch, alpha, rho, (resolution - 1) * 2)
+            The last axis iterates through every bounce integral performed
+            along that field line padded by zeros.
+
+        """
+        pitch, intersect, is_bp, _, _ = _compute_bp_if_given_pitch(
+            pitch, *original, err=True, zeta=zeta, poly_B=poly_B, poly_B_z=poly_B_z
+        )
+        intersect = intersect.reshape(
+            (intersect.shape[0] * intersect.shape[1],) + intersect.shape[2:]
+        )
+        # include knots of spline along with intersection points
+        intersect = jnp.stack(
+            jnp.broadcast_arrays(
+                zeta[:-1],
+                intersect[..., 0],
+                intersect[..., 1],
+                intersect[..., 2],
+                zeta[1:],
+            ),
+            axis=-1,
+        )
+        R = NUM_ROOTS + 2
+        is_bp = stretch_batches(is_bp, NUM_ROOTS, R, fill=False)
+
+        integrand = jnp.nan_to_num(
+            eq.compute(name, grid=grid, override_grid=False, data=data)[name]
+            / (data["B^zeta"] * jnp.sqrt(1 - pitch[:, jnp.newaxis] * data["|B|"]))
+        ).reshape(pitch.size * M * L, resolution)
+
+        # TODO: https://github.com/f0uriest/interpax/issues/19
+        integrand = Akima1DInterpolator(zeta, integrand, axis=-1).c
+
+        integrand = jnp.moveaxis(integrand, 1, -1)
+        assert integrand.shape == (4, pitch.size * M * L, N)
```

Comment:
The assert statements will cause issues with jitting; they are included to make the reviewer's life easier. I'll remove them when the PR gets approved for merge.

Reply:
these should actually be fine since they only depend on static values and array shapes